### ISSUE TITLE
very slow firefox startup

### ISSUE THREAD

**Issue by rcoreilly** [ZotPlus/zotero-better-bibtex#385]
Firefox has been very slow to startup and I finally tracked it down to BBT.  Any ideas?  Anything I can do to debug it, or settings that might be problematic?  Thanks!

**Comment by retorquere** [comment]
Can you go into the BBT prefs, enable extended debugging, close FF (entirely, FF has a habit of keeping windows open), start it up again, and then select "send zotplus report" from the gear menu, and then post the debug ID here. Startup will be even slower than usual, sorry. You can disable it again after the debug report is sent.

**Comment by rcoreilly** [comment]
Here's the debug ID: HXNH9QC2

**Comment by rcoreilly** [comment]
and another one without my proxy errors (which doesn't affect the startup speed) K5G27H6Z

**Comment by retorquere** [comment]
Could you try http://tempsend.com/9A30643902 for me to see if it makes a noticeable difference, and  if not, get a new debug-enabled log from that?

**Comment by rcoreilly** [comment]
that didn't make much of a diff -- perhaps just slightly faster.. here's the new id: UNIRAEJS

**Comment by retorquere** [comment]
OK, that did help a little -- how about http://tempsend.com/CEF7C2C382 ?

**Comment by rcoreilly** [comment]
not obviously better.. N5FXAJDZ

**Comment by retorquere** [comment]
Could you drop /Users/oreilly/Library/Application Support/Firefox/Profiles/fsb46697.default/zotero/betterbibtex.sqlite into https://dbinbox.com/retorquere ?

**Comment by retorquere** [comment]
(note to self: prevent table scan to look up library ID by saving it to the key manager DB)

**Comment by retorquere** [comment]
I won't be able to do much until tomorrow, but I think I have a lead.

**Comment by retorquere** [comment]
Is it correct that you have about 22k items? I'm trying to set up a local simile.

**Comment by rcoreilly** [comment]
yep that is correct

- Randy

**Comment by retorquere** [comment]
OK, I have a working setup and a re-architecting of the cache that should help substantially, but I want to do more tests on startup time and data migration. Probably done before the weekend, currently startup time attributable to BBT for a 32k library is between 1-2 seconds, and initial [performance tests](https://travis-ci.org/ZotPlus/zotero-better-bibtex/jobs/86736675#L564) for export show promising side effects of the change: 1200 item library takes 8.5s for stock exporter, 23s for BBT from an empty cache, but then the cache is filled, and exports take approx 3.3s.

**Comment by retorquere** [comment]
I think I have everything in place, but I have one immodest request: if there is no sensitive data in your references, could I have a private copy of your library? I wouldn't check this in to BBT, so it wouldn't go beyond my system, and I'd destroy it after testing, but I'd really prefer to do a final migration test with a live installation rather than just the synthetic one I've constructed.

If this is OK with you, could you:

1. Exit firefox completely
2. zip up `/Users/oreilly/Library/Application Support/Firefox/Profiles/fsb46697.default/zotero/`
3. Drop the resulting file into https://dbinbox.com/retorquere

Thanks.

**Comment by rcoreilly** [comment]
Sure, no problem to do that — I could also just add you to my shared zotero group if that would be easier?

- Randy

**Comment by retorquere** [comment]
The zip file would also get me a copy of your cache, which I'm looking to
migrate also. I can rebuild the cache after a share+sync, but I'd rather
have your situation as-is.

**Comment by rcoreilly** [comment]
just uploaded — turns out I had ALL my pdf’s sync’d local — many GB’s of files — just deleted those so now it was reasonable to upload.  also double-checked that these PDF's were not the source of the sync delay.

- Randy

**Comment by retorquere** [comment]
The zipfile is still too big -- dbinbox won't accept more than 30MB on my current plan, and it says the file was about 85. This was 85MB without the PDFs? I've sent you a share link for a google drive folder, you can put them there if you wish. Please let me know if that has succeeded.

**Comment by rcoreilly** [comment]
yep, 85mb, without the PDF’s!  uploading to your google drive.

- Randy

**Comment by rcoreilly** [comment]
uploaded..

- Randy

**Comment by retorquere** [comment]
I'll give this backup a shot, but Firefox must still have been running while it was made, as there are `.sqlite-journal` files in the zip. Could you give it another go? The database state [is uncertain](https://www.sqlite.org/howtocorrupt.html) when backups are made on a running system.

**Comment by rcoreilly** [comment]
new zotero2.zip now uploading, created when firefox was closed.  sorry about that!

- Randy

**Comment by retorquere** [comment]
46s to export your library. Not bad!

**Comment by retorquere** [comment]
(after the initial cache priming, which took ages)

**Comment by retorquere** [comment]
And a DB load time of 2.7 seconds. Still a few more things to fix -- the serialization cache is somethimes dropped, and I haven't yet figured out why.

**Comment by retorquere** [comment]
Got it. I am going to do a little more testing on this, but I'm very satisfied with the results, and all the existing tests in the test suite pass. The change will also ease the transition to Zotero 5.0 as a bonus, so it's definitely going in. If you're feeling adventurous:

1. Make a backup of `/Users/oreilly/Library/Application Support/Firefox/Profiles/fsb46697.default/zotero/`. The upgrade makes its own backups, but it doesn't hurt to have one extra.
2. Install the [pre-release](http://tempsend.com/03D7E1AC90)
3. Enjoy fast startup & exports

Performance is actually [substantially better](https://travis-ci.org/ZotPlus/zotero-better-bibtex/jobs/87369374#L566) than it was before; with a primed cache, BBT is now around 5 times as fast as the stock BibLaTeX exporter (although the first run to prime the cache is about 3 times as slow as the stock exporter). 

First startup will take a little while as it migrates the BBT database; it will not be as long as the startup time you have had so far, and after migration I'm getting consistent 1-second BBT DB loads on your library.

The cache can be primed as you go, but if you want to get it over with, export the whole library and go do something productive. Your library took about 40 minutes on my laptop the first time, but approx 30 seconds after that.

**Comment by retorquere** [comment]
Oh and you'll be the first to be able to marvel at the newly included cache activity report at http://localhost:23119/better-bibtex/cacheActivity (if you have HTTP export on). This is still under development, but immediately after an export it gives you some idea of what's going on.

**Comment by rcoreilly** [comment]
exciting! I’ll download the pre-release and let you know how it goes!  Thanks!!!

- Randy

**Comment by retorquere** [comment]
Super, I really appreciate your help in getting this nailed down. You don't
have to remember that link, it's included in the debug section of the BBT
prefs.

**Comment by rcoreilly** [comment]
Everything worked as advertised — took FOREVER to do the first export but second seemed faster even than 30sec.  And startup is essentially instantaneous!  The only thing that didn’t work for me was the cache browser.  I got this error:

Firefox can't establish a connection to the server at localhost:23119.

I turned off my proxy (even though it has an exception for localhost) just to make sure it wasn’t that..

Thanks again!!
- Randy

**Comment by retorquere** [comment]
*grin* forever sounds about right. For the cache activity, go to Prefs/BBT/Import-Export and enable "export by HTTP". I'll add an indication to the debug pane on the status.

**Comment by retorquere** [comment]
BTW a relevant test would be to change an item that doesn't have a pinned key (so the key would change), close FF, start FF again, export that reference twice, and see if it has the expected key both times. It does for me, but I'd love to have your confirmation.

**Comment by retorquere** [comment]
http://tempsend.com/893B87867A explains how to enable cache activity reports in the debug pane itself.

**Comment by retorquere** [comment]
the 385 branch has been merged into master so any further activity will happen there.

**Comment by retorquere** [comment]
It should be illegal to be this happy about such things, but I love the performance work that has resulted from this, plus: pretty charts! I've changed to a stepped area chart (see http://tempsend.com/08B6C67AE2), but I'd love to get your feedback on which of [these](https://developers.google.com/chart/interactive/docs/gallery) would be best to visualize these things.

**Comment by retorquere** [comment]
(the cache activity link has changed BTW; the debug pane has been updated accordingly)

**Comment by rcoreilly** [comment]
It passed your test of changing an item, restarting FF, etc.  And I think the stacked area chart makes the most sense for that data.  When will you be releasing the official update?  My lab members are eager to use it!  Thanks,

- Randy

**Comment by retorquere** [comment]
It's out -- enjoy!

**Comment by smatthie** [comment]
Wow! Whatever you did, it's an enormous speedup. I just came back from holidays and got the update. I was actually about to open an issue because the export took something like 10-15 minutes, blocking FF completely, but after reading this thread I checked for the update, restarted FF, and each export now takes a few seconds - I didn't even notice that it was doing it. The automatic export is quite important for me because I synchronise the bibtex with a table so that I can access my literature offline.

Perfect and many many thanks for your work!

**Comment by retorquere** [comment]
Thank you for chiming in. Working on bbt is a great pleasure, and input from users always helps to make it better. 

Wrt what I did: generating the translator output requirsd two transformations per reference, one from database tables to javascript object, one from javascript object to bibtex (or any other output format). I now maintain two caches; if nothing has changed I go from reference ID straight to bibtex, and if you re-export the same reference to a new bbt format it will at least skip the first step. 

I've also moved all bbt data (including the caches)  to an in-memory javascript database, as the overhead of lookups in the disk-based database incurred two major penalties; the lookup itself, but also the translation from C (in which the db was written) to javascript. This database is saved (if changed) whenever your computer goes idle, or when 10 seconds have passed, whichever comes first. 

Bottom line, a whole heap of things don't get repeated unnecessarily, and when it is necessary, it is done from memory rather than disk whenever possible. The bbt translators are a lot slower than the stock translators by a factor of 6 iirc (but then the stock translators are no match for the bbt translators in terms of bibtex output), but after the first export of a reference which primes the cache, BBT is faster than stock by a factor of 5. Given the fact that references change very infrequently when compared to the number of times each is exported, a clear win. 

The cache requires cooperation by the translator so it isn't applicable to non-bbt translators, and it is only active when you don't export attachments, so for exports with attachments, the stock exporter will always be faster.
